platform/linux-generic/pktio/dpdk.c

```diff
@@ -986,21 +991,34 @@ static inline int pkt_to_mbuf_zero(pktio_entry_t *pktio_entry,
 			if (odp_unlikely(chksum_enabled))
 				pkt_set_ol_tx(pktout_cfg, pktout_capa, pkt_hdr,
 					      mbuf, odp_packet_data(pkt));
+
+			pkt_info[i].is_copied = false;
 		} else {
 			int dummy_idx = 0;
 
 			/* Fall back to packet copy */
 			if (odp_unlikely(pkt_to_mbuf(pktio_entry, &mbuf,
 						     &pkt, 1, &dummy_idx) != 1))
 				goto fail;
-			(*copy_count)++;
+
+			pkt_info[i].is_copied = true;
 		}
+
 		mbuf_table[i] = mbuf;
 
 		if (odp_unlikely(tx_ts_enabled)) {
 			if (odp_unlikely(*tx_ts_idx == 0 && pkt_hdr->p.flags.ts_set))
 				*tx_ts_idx = i + 1;
 		}
+
+		if (odp_unlikely(tx_compl_enabled && pkt_hdr->p.flags.tx_compl &&
+				 pkt_hdr->tx_compl_mode == ODP_PACKET_TX_COMPL_ALL)) {
+			pkt_info[i].tx_compl_q = pkt_hdr->tx_compl_queue;
+			pkt_info[i].user_ptr = pkt_hdr->user_ptr;
+		} else {
+			pkt_info[i].tx_compl_q = ODP_QUEUE_INVALID;
+			pkt_info[i].user_ptr = NULL;
```

Comment:
If there would be an additional output argument (e.g. postprocess), where you would return the number of compl events/copies, this branch could be skipped. All writing could be done in the unlikely branch.